pkg/controllers/subnetset/subnetset_webhook.go

```diff
@@ -0,0 +1,90 @@
+/* Copyright © 2023 VMware, Inc. All Rights Reserved.
+   SPDX-License-Identifier: Apache-2.0 */
+
+package subnetset
+
+import (
+	"context"
+	"net/http"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+)
+
+// log is for logging in this package.
+var subnetsetlog = logf.Log.WithName("subnetset-webhook")
+
+var NSXOperatorSA = "system:serviceaccount:vmware-system-nsx:nsx-operator-svc-account"
+
+// Create validator instead of using the existing one in controller-runtime because the existing one can't
+// inspect admission.Request in Handle function.
+
+//+kubebuilder:webhook:path=/validate-nsx-vmware-com-v1alpha1-subnetset,mutating=false,failurePolicy=fail,sideEffects=None,groups=nsx.vmware.com.nsx.vmware.com,resources=subnetsets,verbs=create;update,versions=v1alpha1,name=vsubnetset.kb.io,admissionReviewVersions=v1
+
+type SubnetSetValidator struct {
+	Client  client.Client
+	decoder *admission.Decoder
+}
+
+func isDefaultSubnetSet(s *v1alpha1.SubnetSet) bool {
+	if _, ok := s.Labels[common.LabelDefaultSubnetSet]; ok {
+		return true
+	}
+	if s.Name == common.DefaultVMSubnetSet || s.Name == common.DefaultPodSubnetSet {
+		return true
+	}
+	return false
+}
+
+// Handle handles admission requests.
+func (v *SubnetSetValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	subnetSet := &v1alpha1.SubnetSet{}
+	if req.Operation == admissionv1.Delete {
+		err := v.decoder.DecodeRaw(req.OldObject, subnetSet)
+		if err != nil {
+			subnetsetlog.Error(err, "error while decoding SubnetSet", "SubnetSet", req.Namespace+"/"+req.Name)
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+	} else {
+		err := v.decoder.Decode(req, subnetSet)
+		if err != nil {
+			subnetsetlog.Error(err, "error while decoding SubnetSet", "SubnetSet", req.Namespace+"/"+req.Name)
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+	}
+	if !isDefaultSubnetSet(subnetSet) {
+		return admission.Allowed("")
+	}
+	subnetsetlog.Info("request user-info", "name", req.UserInfo.Username)
+	switch req.Operation {
+	case admissionv1.Create:
+		if req.UserInfo.Username == NSXOperatorSA {
+			return admission.Allowed("")
+		}
+		return admission.Denied("default SubnetSet only can be created by nsx-operator")
+	case admissionv1.Update:
+		if req.UserInfo.Username == NSXOperatorSA {
+			// nsx-operator needs to update tags/default-values/status for default SubnetSet.
```

Comment:
can we also forbid nsx-operator user from "add, modify or delete" the default subnetset label?

Reply:
> can we also forbid nsx-operator user from "add, modify or delete" the default subnetset label?

Sure we can. But a minor question, do we really need to do this? As we know that nsx-operator will never perform these operations.

Reply:
Updated according to our previous offline discussion.